zcash_client_sqlite/src/wallet.rs

```diff
@@ -140,82 +140,71 @@ pub(crate) const BLOCK_SAPLING_FRONTIER_ABSENT: &[u8] = &[0x0];
+        (0, Some(seed_fp), Some(account_index)) => Ok(AccountType::Derived {
+            seed_fingerprint: HdSeedFingerprint::from_bytes(seed_fp),
+            account_index: zip32::AccountId::try_from(account_index).map_err(|_| {
+                SqliteClientError::CorruptedData(
+                    "ZIP-32 account ID from wallet DB is out of range.".to_string(),
+                )
+            })?,
+        }),
+        (1, None, None) => Ok(AccountType::Imported),
+        (0, None, None) | (1, Some(_), Some(_)) => Err(SqliteClientError::CorruptedData(
+            "Wallet DB account_type constraint violated".to_string(),
+        )),
+        (_, _, _) => Err(SqliteClientError::CorruptedData(
+            "Unrecognized account_type".to_string(),
+        )),
     }
 }
 
-impl From<AccountType> for u32 {
-    fn from(value: AccountType) -> Self {
-        match value {
-            AccountType::Zip32 => 0,
-            AccountType::Imported => 1,
-        }
+fn account_kind_code(value: AccountType) -> u32 {
+    match value {
+        AccountType::Derived { .. } => 0,
+        AccountType::Imported => 1,
     }
 }
 
-/// Describes the key inputs and UFVK for an account that was derived from a ZIP-32 HD seed and account index.
+/// The viewing key that an [`Account`] has available to it.
 #[derive(Debug, Clone)]
-pub(crate) struct HdSeedAccount(
-    HdSeedFingerprint,
-    zip32::AccountId,
-    Box<UnifiedFullViewingKey>,
-);
-
-impl HdSeedAccount {
-    pub fn new(
-        hd_seed_fingerprint: HdSeedFingerprint,
-        account_index: zip32::AccountId,
-        ufvk: UnifiedFullViewingKey,
-    ) -> Self {
-        Self(hd_seed_fingerprint, account_index, Box::new(ufvk))
-    }
-
-    /// Returns the HD seed fingerprint for this account.
-    pub fn hd_seed_fingerprint(&self) -> &HdSeedFingerprint {
-        &self.0
-    }
-
-    /// Returns the ZIP-32 account index for this account.
-    pub fn account_index(&self) -> zip32::AccountId {
-        self.1
-    }
-
-    /// Returns the Unified Full Viewing Key for this account.
-    pub fn ufvk(&self) -> &UnifiedFullViewingKey {
-        &self.2
-    }
-}
-
-/// Represents an arbitrary account for which the seed and ZIP-32 account ID are not known
-/// and may not have been involved in creating this account.
-#[derive(Debug, Clone)]
-pub(crate) enum ImportedAccount {
-    /// An account that was imported via its full viewing key.
+pub(crate) enum ViewingKey {
+    /// A full viewing key.
+    ///
+    /// This is available to derived accounts, as well as accounts directly imported as
+    /// full viewing keys.
     Full(Box<UnifiedFullViewingKey>),
-    /// An account that was imported via its incoming viewing key.
+
+    /// An incoming viewing key.
+    ///
+    /// Accounts that have this kind of viewing key cannot be used in wallet contexts,
+    /// because they are unable to maintain an accurate balance.
     Incoming(Uivk),
```

Comment:
Note that this will change in https://github.com/zcash/librustzcash/pull/1245

Reply:
Yep that's fine; this is still a private enum.